File: leetcode/217.py
"""
Runtime: 112 ms, faster than 90.52% of Python3 online submissions for Contains Duplicate.
Memory Usage: 21.6 MB, less than 22.02% of Python3 online submissions for Contains Duplicate.
"""
from collections import Counter


class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        return any([v >= 2 for v in Counter(nums).values()])

File: leetcode/217.cpp
// Runtime: 101 ms
// Memory Usage: 72 MB

/*
  O(n) to fill the counter and search.
*/
#include <map>

class Solution {
 public:
  bool containsDuplicate(vector<int>& nums) {
    // count occurrences
    std::unordered_map<int, int> counter;
    for (size_t i = 0; i < nums.size(); ++i) {
      counter[nums[i]] += 1;
    }

    // check for duplicates
    for (auto pair : counter) {
      if (pair.second > 1) {
        return true;
      }
    }

    return false;
  }
};
